### ISSUE TITLE
Pump behavior improvements

### ISSUE THREAD

**Issue by jshipley** [Rearth/Oritech#51]
### Describe the feature you'd like

Here are some things that the pump should do:

* If pump or pump trunk is broken, all pump trunk blocks below should also be broken
* If pump trunk is broken, pump should lower a new pump trunk before continuing to pump fluid
* When pump is finished pumping available fluids, it should retract the pump trunk
* There should be some way to restart the fluid search. For instance, if a player adds a water source block under the pump to pump infinite water, the pump will never recognize it unless it is broken and placed again.
* As a continuation, it might be nice to have a custom GUI for the pump that says what state the pump is in (searching, pumping, low power, inventory full, etc.), and how many buckets of fluid it is expecting to pump. The GUI could also allow players to fill buckets from the pump
* Rather than just treating water as infinite, the pump should check if a fluid source is infinite. It could be a 1x1 or 2x1 water which is infinite, or a 3x3 lava with the lavaSourceConversion game rule set to true, or some other modded fluid that is set to be infinite.
* For some reason, pump seems to always face the same direction, regardless of the direction the player is facing when placing it

**Comment by jshipley** [comment]
The pump should also work if it's directly above the water and doesn't have space to expand the trunk.

**Comment by Rearth** [comment]
Hi, thanks for the suggestions. In response I just pushed a few changes to the pump on the main branch. This include the change that all pump blocks are now also broken when the pump is broken. I also added the missing facing property so it now properly rotates when placed.

I decided against a GUI for now, as I wanted to skip the extra work. Once I have a better GUI/Networking system in place I'll probably revisit this. Instead I just added a status message when right clicked. This should also work for now.

Retracting the trunk automatically when finished, restarting fluid search, stopping when trunk is broken is out of the scope for now. Just break and build it again if that's really needed. All those are good suggestion but I currently don't want to extend the pump logic too much, as those topics are pretty low priority for me.

Actually checking if a fluid is infinite is something I tried, but there's no direct support or APIs for that, so modded fluids won't ever work. Atleast I couldnt find anything that checks if a fluid is infinite.

**Comment by jshipley** [comment]
I remembered seeing the isInfinite() method in the FlowableFluid class, but it looks like that just returns whether or not semi-adjacent source blocks will combine into new source blocks. Lava/Water, for example, just return whether their source conversion game rules are set to true.

You'd have to combine that with a check to see if there's another source block within two moves of it that can combine into a new source block. It's already kind of an edge case, and the current handling is good enough for now.

I'd be interested to see how the pump behaves if it tries to pump a body of infinite milk, though. Does it just visit every source in the pool, drain it, and then move on and consider the block drained even though it was soon refilled? Then finish "draining" the entire pool and think it has nothing left to pump even though the entire pool is still there?

**Comment by Rearth** [comment]
Pretty sure milk isnt an actual liquid in minecraft, right?

**Comment by jshipley** [comment]
It's not, but its a common modded liquid that can be infinite. I would have said lava, but that flows slowly enough that it wouldn't be a very good test.

--
Jeff Shipley

**Comment by jshipley** [comment]
Milk was a bust. The easiest way to add milk (tropheusj's Milk+ library) hasn't been updated to 1.21 yet.

I ended up testing with oil (by returning true from MineralFluid.isInfinite()), and it worked exactly like I thought it would. The pump pulled every block of fluid, which was replaced by adjacent fluid blocks. Once it had processed every block it thought it was done, even though the entire pool of fluid was still there. Lava (with lavaSourceConversion set to true) also behaved (in the overworld, at least) like I expected: it was sucked up faster than the source conversion was able to replace it. That's just a 3x5x1 pool of lava, though. A larger pool might have better behavior. Besides, even if lava is infinite, I don't think that sucking it up faster than it can convert is a problem or would be really unexpected.

Also, the pump facing property is working great, but it kind of feels like the pump is always placed sideways. Not a major concern, and will be a complete non-issue when the pump model is eventually replaced.

I might still do a PR for better generic infinite fluid handling at some point, but as far as I know most of the mods with modded fluids aren't even available on 1.21 yet and even then I'm not aware of many that are actually infinite

I'd still like to see an idle pump reinitialized when the player right-clicks on it, but breaking the pump and placing it again is good enough. The main situation I can think of where this would be needed is when players place a pump them put an infinite water source below it.

Thanks for the improvements!
